File: DsaWithCpp/Recursion/FindFirstAndLastOccurance.cpp
#include <iostream>
using namespace std;

int first = -1, last = -1;

void findFirstAndLastOccurance(string str, int i, char ch);

int main()
{
    string str;
    char ch;
    cout << "Enter the String -> ";
    cin >> str;
    cout << "Enter the Character you want to track -> ";
    cin >> ch;
    findFirstAndLastOccurance(str, 0, ch);
}
void findFirstAndLastOccurance(string str, int i, char ch)
{
    if (i == str.length() - 1)
    {
        printf("First & Last Occurance of character %c are %d and %d respectively.", ch, first, last);
        return;
    }
    if (ch == str.at(i))
    {
        if (first == -1)
            first = last = i;
        else
            last = i;
    }
    findFirstAndLastOccurance(str, i + 1, ch);
}

File: dsa_with_python/recursion/find_first_and_last_occurance.py
first = last = -1


def find_first_and_last_occurance(str: str, i: int, char: str):
    global first, last
    if i == len(str) - 1:
        print(f"First & Last Occurance of character {char} are {first} and {last} respectively.")
        return
    if char == str[i]:
        if first == -1:
            first = last = i
        else:
            last = i
    find_first_and_last_occurance(str, i + 1, char)


str = input("Enter the String -> ")
char = input("Enter the Character you want to track -> ")
find_first_and_last_occurance(str, 0, char)
